File: coding/src/com/algos/core/li19_bit/revised/Basics.java
package  com.algos.core.li19_bit.revised;

public class Basics {

	public static void main(String[] args) {
		Basics obj = new Basics();

		// Binary iteratively
		obj.printBinaryIteratively(4);
		
		// Binary recursively
		obj.printBinaryRecursively(4);
		System.out.println();
		
		// Binary recursively with right shift	
		obj.printBinaryRecursivelyWithRightShift(4);
		System.out.println();
		
		// Get the rightmost set bit
		int x = 100;
		System.out.println(Integer.toBinaryString(x));	// 1100100
		System.out.println(x&~(x-1));	// 4 -> 0000100
	}
	private void printBinaryIteratively(int n){
		StringBuffer sb = new StringBuffer();
		while(n > 0){
			sb.append(n%2);
			n = n/2;
		}
		System.out.println(sb.reverse().toString());
	}
	private void printBinaryRecursively(int n){
		if(n > 0){
			printBinaryRecursively(n/2);
		}
		System.out.print(n%2);
	}
	private void printBinaryRecursivelyWithRightShift(int n){
		if(n > 0){
			printBinaryRecursively(n>>1);
		}
		System.out.print(n&1);
	}
}

File: dsa/topics/19_bit/python/basics.py
def print_binary_iteratively(n):
    sb = []
    while n > 0:
        sb.append(str(n % 2))
        n = n // 2
    print("".join(reversed(sb)))


def print_binary_recursively(n):
    if n > 0:
        print_binary_recursively(n // 2)
    print(n % 2, end="")


def print_binary_recursively_with_right_shift(n):
    if n > 0:
        print_binary_recursively(n >> 1)
    print(n & 1, end="")


# ------------------ Test (Same as Java main) ------------------

# Binary iteratively
print_binary_iteratively(4)

# Binary recursively
print_binary_recursively(4)
print()

# Binary recursively with right shift
print_binary_recursively_with_right_shift(4)
print()

# Get the rightmost set bit
x = 100
print(bin(x)[2:])        # 1100100
print(x & ~(x - 1))      # 4 -> 0000100
